docs/basics/upgradeability.md

```diff
@@ -0,0 +1,183 @@
+        ink_env::call::build_call::<ink_env::DefaultEnvironment>()
+            .call_type(
+                Call::new()
+                    .callee(self.forward_to)
+                    .transferred_value(self.env().transferred_value())
+                    .gas_limit(0),
+            )
+            .call_flags(
+                ink_env::CallFlags::default()
+                    .set_forward_input(true)
+                    .set_tail_call(true),
+            )
+            .fire()
+            .unwrap_or_else(|err| {
+                panic!(
+                    "cross-contract call to {:?} failed due to {:?}",
+                    self.forward_to, err
+                )
+            });
+        unreachable!(
+            "the forwarded call will never return since `tail_call` was set"
+        );
+    }
+}
+```
+
+Using this pattern, you can introduce other message to your proxy contract.
+Any messages that are not matched on the proxy contract will be forwarded the specified contract address.
+
+# Updating contract code directly
+
+Following [Substrate's runtime upgradeability](https://docs.substrate.io/tutorials/get-started/forkless-upgrade/) philosophy, ink! also supports an easy way to update your contract code via the special function `set_code_hash()`
+
+## Properties
+
+- Updates the contract code using set_code_hash. This effectively replaces the code which is executed for the contract address.
+- The other contract needs to be deployed on-chain.
+- State is stored in the storage of the originally instantiated contract.
+
+Just add the following function to the contract your want to upgrade in future.
+
+```rust 
+/// Modifies the code which is used to execute calls to this contract address (`AccountId`).
+///
+/// We use this to upgrade the contract logic. We don't do any authorization here, any caller
+/// can execute this method. In a production contract you would do some authorization here.
+#[ink(message)]
+pub fn set_code(&mut self, code_hash: [u8; 32]) {
+    ink_env::set_code_hash(&code_hash).unwrap_or_else(|err| {
+        panic!(
+            "Failed to `set_code_hash` to {:?} due to {:?}",
+            code_hash, err
+        )
+    });
+    ink_env::debug_println!("Switched code hash to {:?}.", code_hash);
+}
+```
+
+## Storage compatibility
+
+It is developer's responsibility to ensure that the new contract's storage is compatible with the old version.
+
+**You should not change the order in which the contract state variables are declared, nor their type!**
+
+Violating the restriction will not prevent a successful compilation,
+but will result in **the mix-up of values** or **failure to read the storage correctly**.
+This can be a result of severe errors in the application utilizing the contract.
+
+If the storage of your contract looks like this:
+```rust
+use ink_lang as ink;
+#[ink(storage)]
+pub struct YourContract {
+    x: u32,
+    y: bool,
+}
+```
+The procedures listed below will make it ***invalid***:
```

Comment:
Not sure how this looks like rendered, but I'd still had a bit of space around
the codeblocks and commentary to make it easier to read in the source